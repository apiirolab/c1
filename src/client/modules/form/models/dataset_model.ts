export interface DataSet {
  getValue(name: string): any;
  parseValue(name: string, value: any): void;
  getError(name: string): any;
  getStringValue(name: string): any;
  setStringValue(name: string, value: any): void;
  isExpression(key: string): boolean;
}
