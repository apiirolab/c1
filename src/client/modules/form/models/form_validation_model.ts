export interface FormValidationDao {}

export interface FormValidation {}
