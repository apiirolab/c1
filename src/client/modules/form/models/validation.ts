import i18n from 'es2015-i18n-tag';

export function requiredValidator(value: string | number) {
  return value === '' || value == null ? i18n`This field is required!` : '';
}

export function regExValidator(reg: RegExp, format?: string) {
  return (value: string) => {
    if (reg.exec(value)) {
      return '';
    }
    if (format) {
      return i18n`Expecting format: ${format}`;
    }
    return i18n`Unexpected format`;
  };
}

const intReg = /^(\+|\-)?\d+$/;
function isInt(n: string) {
  return intReg.test(n);
}

const positiveIntReg = /^(\+)?\d+$/;
function isPositiveInt(n: string) {
  return positiveIntReg.test(n);
}

const nonZeroIntReg = /^(\+)?[1-9]\d*$/;
function isNonZeroInt(n: string) {
  return nonZeroIntReg.test(n);
}

const floatReg = /^(\+|\-)?\d*\.?\d+$/;
function isFloat(n: string) {
  return floatReg.test(n);
}

const positiveFloatReg = /^(\+)?\d*\.?\d+$/;
function isPositiveFloat(n: string) {
  return positiveFloatReg.test(n);
}

const nonZeroFloatReg = /^(\+)?[1-9]\d*\.?\d+$/;
function isNonZeroFloat(n: string) {
  return nonZeroFloatReg.test(n);
}

const emailReg = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
function isEmail(n: string) {
  return emailReg.test(n);
}

export function emailValidator(value: string) {
  return value == null || value === '' || isEmail(value) ? '' : i18n`Email has incorrect format!`;
}

export function intPositiveValidator(value: number | string) {
  return value == null || value === '' || isPositiveInt(value.toString())
    ? ''
    : i18n`Expected positive integer value`;
}

export function intValidator(value: number | string) {
  return value == null || value === '' || isInt(value.toString())
    ? ''
    : i18n`Expected integer value`;
}

export function intNonZeroValidator(value: number | string) {
  return value == null || value === '' || isNonZeroInt(value.toString())
    ? ''
    : i18n`Expected non-zero integer value`;
}

export function floatValidator(value: number | string) {
  return value == null || value === '' || isFloat(value.toString())
    ? ''
    : i18n`Expected float value`;
}

export function floatPositiveValidator(value: number | string) {
  return value == null || value === '' || isPositiveFloat(value.toString())
    ? ''
    : i18n`Expected positive float value`;
}

export function floatNonZeroValidator(value: number | string) {
  return value == null || value === '' || isNonZeroFloat(value.toString())
    ? ''
    : i18n`Expected non-zero float value`;
}

export function lengthValidator(length: number, message?: string) {
  return (value: number | string) => {
    return value == null || value === '' || value.toString().length >= length
      ? ''
      : message || i18n`Too short! Minimum ${length} characters`;
  };
}

export function equalityValidator(comparer: () => string | string, message?: string) {
  return (value: string) => {
    let val1 = value ? value.toString() : value.toString();
    let val2 = typeof comparer === 'function' ? comparer() : comparer;

    return val1 === val2 ? '' : message || i18n`Value ${val1} does not match ${val2}`;
  };
}
