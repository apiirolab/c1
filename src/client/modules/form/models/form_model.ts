import { FormValidation } from './form_validation_model';
import { types, ISimpleType } from 'mobx-state-tree';

function formItemSort(a: Corpix.Collections.FormElementDao, b: Corpix.Collections.FormElementDao) {
  return a.row < b.row
    ? -1
    : a.row > b.row
      ? 1
      : a.column < b.column
        ? -1
        : a.column > b.column
          ? 1
          : 0;
}

function mstTypeFactory(type: string): ISimpleType<any> {
  switch (type) {
    case 'string':
      return types.string;
    case 'int':
      return types.number;
  }
  return types.string;
}

export class FormModel {
  id: string;
  name: string;
  description: string;
  elements: Corpix.Collections.FormElementDao[];
  validations: FormValidation[];

  static buildMST(
    descriptors: Corpix.Collections.DataDescriptorDao[],
    dataArray: Array<{ name: string; value: any }>
  ) {
    // data initialisation
    const data: any = {};
    for (let element of dataArray) {
      data[element.name] = element.value;
    }

    // prepare model and views
    const mstDefinition: { [index: string]: string } = {};
    const viewDefinition: { [index: string]: Function } = {};

    for (let desc of descriptors) {
      if (desc.expression) {
        // @ts-ignore // we need this as we need self
        viewDefinition[desc.name] = function(self: any) {
          return eval(desc.expression);
        };
      } else {
        // view
        mstDefinition[desc.name] = desc.defaultValue
          ? desc.defaultValue
          : types.maybe(mstTypeFactory(desc.type));
      }
    }

    // build tree
    const mst = types
      .model('Store', mstDefinition)
      .views(() => viewDefinition)
      .actions(self => ({
        getValue(key: string) {
          return self[key];
        },
        setValue(key: string, value: any) {
          self[key] = value;
        }
      }));

    return mst.create(data);
  }

  constructor(form: Corpix.Collections.FormDao) {
    this.id = form.id;
    this.name = form.name;
    this.description = form.description;
    this.validations = form.validations;

    // add and sort elements
    this.elements = form.elements;
    this.elements.sort(formItemSort);
  }
}
