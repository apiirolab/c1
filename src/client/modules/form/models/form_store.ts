import { IObservableArray, toJS } from 'mobx';
import { types } from 'mobx-state-tree';
import { UndoManager } from 'mst-middlewares';

import { QueryTypes } from 'data/client';
import { DataDescriptor, DataSet } from './form_model';

export type IValidator = (input: string) => string;

export type IFormStore = typeof FormStore.Type;

export type ValidationResult = {
  valid: number;
  invalid: number;
  total: number;
};

function strip(obj: any) {
  if (typeof obj === 'object') {
    if (obj.errors) {
      delete obj.errors;
      delete obj.validators;
      delete obj.strings;
      delete obj.descriptors;
      delete obj.arrays;
      delete obj.objects;
    }

    for (let key of Object.getOwnPropertyNames(obj)) {
      let item = obj[key];
      strip(item);
    }
  }

  if (Array.isArray(obj)) {
    for (let item of obj) {
      strip(item);
    }
  }
  return obj;
}

export function isRequired(descriptor: QueryTypes.DataDescriptor) {
  return (
    descriptor.validators &&
    descriptor.validators.length &&
    descriptor.validators.some(d => d.name === QueryTypes.ValidatorFunction.RequiredValidator)
  );
}

export interface ListValue {
  [index: string]: string;
  text: string;
  value: string;
}

export const FormStore = types
  .model({
    dirty: false,
    errors: types.map(types.string),
    strings: types.map(types.string)
  })
  .volatile(self => ({
    arrays: [],
    objects: [],
    // tslint:disable-next-line:no-object-literal-type-assertion
    validators: {} as { [name: string]: IValidator[] },
    // tslint:disable-next-line:no-object-literal-type-assertion
    descriptors: {} as { [index: string]: DataDescriptor },
    helpers: {
      getDescriptor(key: string) {
        if (!self.descriptors[key]) {
          throw new Error('Descriptor does not exist: ' + key);
        }
        return self.descriptors[key];
      },
      isExpression(key: string) {
        return !!(key && this.getDescriptor(key).expression);
      }
    }
  }))
  .actions(function(self) {
    // let validators: values: { [name: string]: IValidator[] } = new Map<string, IValidator[]>();

    function setValue(key: string, value: any) {
      if ((self as any)[key] !== value) {
        (self as any)[key] = value;
        self.dirty = true;
      }
    }

    return {
      saved() {
        self.dirty = false;
      },
      toJS() {
        return strip(toJS(self));
      },

      getList(key: string): ListValue[] {
        return (self as any)[key];
      },
      validateField(name: string, value: string) {
        if (!self.validators[name]) {
          return true;
        }

        self.errors.set(name, '');

        for (let v of self.validators[name]) {
          if (v) {
            let result = v(value);
            if (result) {
              self.errors.set(name, result);
              return false;
            }
          }
        }

        return true;
      },
      // setValidators(values: { [name: string]: IValidator[] }) {
      //   for (let key of Object.getOwnPropertyNames(values)) {
      //     self.validators[key] = values[key];
      //   }
      // },
      setError(item: string, value: string) {
        self.errors.set(item, value);
        if (value) {
          self.errors.set(item, value);
        } else {
          self.errors.delete(item);
        }
      },
      setItem(item: string, value: string): void {
        (self as any)[item] = value;
        (self as IFormStore).validateField(item, value);
      },
      addRow(key: string) {
        ((self as any)[key] as any[]).push({});
      },
      removeRow(key: string, index: number) {
        ((self as any)[key] as any[]).splice(index);
      },
      removeRowData<T>(key: string, data: T) {
        ((self as any)[key] as IObservableArray<T>).remove(data);
      },
      parseValue(key: string, value: any) {
        const descriptor = self.descriptors[key];
        switch (descriptor.type) {
          case 'Int':
            setValue(key, parseInt(value || 0, 10));
            break;
          case 'Float':
            setValue(key, parseFloat(value || 0));
            break;
          case 'Boolean':
            setValue(key, value === true || value === 'true' || value === 'True');
            break;
          default:
            setValue(key, value);
        }
      }
    };
  })
  .actions(self => ({
    validateWithReport(): ValidationResult {
      let total = 0;
      let valid = 0;

      // validate self

      for (let key of Object.getOwnPropertyNames(self.validators)) {
        total += 1;
        if (self.validateField(key, (self as any)[key])) {
          valid += 1;
        }
      }

      // validate objects

      for (let key of self.objects) {
        let obj = (self as IFormStore).getValue(key) as DataSet;
        let result = obj.validateWithReport();
        total += result.total;
        valid = result.valid;
      }

      // validate arrays

      for (let key of self.arrays) {
        let obj = (self as IFormStore).getValue(key) as DataSet[];

        for (let row of obj) {
          let result = row.validateWithReport();
          total += result.total;
          valid = result.valid;
        }
      }

      // create report

      return {
        total,
        valid,
        invalid: total - valid
      };
    }
  }))
  .actions(self => ({
    validate() {
      return self.validateWithReport().invalid === 0;
    },
    setStringValue(key: string, value: any, validate = true) {
      self.strings.set(key, value);

      if (validate) {
        let isOK = self.validateField(key, value);

        if (isOK) {
          self.parseValue(key, value);
        }
      }
      // self.setError(key, error);
    }
  }))
  .views(self => ({
    getValue(item: string): any {
      return (self as any)[item];
    },
    getStringValue(key: string): string {
      const descriptor = self.helpers.getDescriptor(key);
      // expressions are evaluated on the fly
      if (descriptor.expression) {
        // @ts-ignore
        return self[key];
      }

      if (descriptor.type === QueryTypes.DataType.Object) {
        return '<ERROR! Rendering Object as TEXT>';
      }

      // init
      if (!self.strings.has(key)) {
        self.setStringValue(
          key,
          (self as any)[key] == null ? '' : (self as any)[key].toString(),
          false
        );
      }
      return self.strings.get(key); // || this.getDescriptor(key).defaultValue || '';
    },
    getError(item: string): string {
      return self.errors.get(item);
    }
  }));

// export const FormStore = types
//   .model({
//     errors: types.map(types.string)
//   })
//   .actions(self => ({
//     setItem(item: string, value: string): void {
//       (self as any)[item] = value;
//     },
//     getItem(item: string): string {
//       return (self as any)[item];
//     },
//     getError(item: string): string {
//       return self.errors.get(item);
//     }
//   }));
