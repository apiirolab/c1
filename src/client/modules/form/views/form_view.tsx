import * as React from 'react';

import { Grid, SemanticWIDTHSNUMBER, Input } from 'semantic-ui-react';

import { groupByArray } from 'shared/helpers';

import { DataSet } from '../models/dataset_model';
import { FormModel } from '../models/form_model';
import { FormControl } from './input_view';

interface Props {
  data: DataSet;
  form: FormModel;
}

export class FormView extends React.Component<Props> {
  lastRow = -1;
  lastColumn = -1;

  renderControl(control: Corpix.Entities.FormElement, dataSet: DataSet) {
    const formElement = control as Corpix.Entities.FormControl;
    let formControl;

    switch (control.control) {
      case 'input':
        formControl = Input;
        break;
      default:
        throw new Error('Not implemented: ' + control.control);
    }

    return (
      <FormControl
        control={formControl}
        owner={dataSet}
        name={formElement.source}
        label={formElement.label}
        inputLabel={formElement.inputLabel}
      />
    );
  }

  renderColumn(control: Corpix.Entities.FormElement) {
    if (control.row != this.lastRow) {
      this.lastRow = control.row;
      this.lastColumn = 0;
    }
    // we initialise all columns and add missing ones in between
    let columns = [];

    // insert missing start column
    if (control.column > this.lastColumn) {
      columns.push(
        <Grid.Column
          key={this.lastColumn}
          width={(control.column - this.lastColumn) as SemanticWIDTHSNUMBER}
        />
      );
    }

    columns.push(
      <Grid.Column key={control.column} width={control.width as SemanticWIDTHSNUMBER}>
        {this.renderControl(control, this.props.data)}
      </Grid.Column>
    );

    this.lastColumn = control.column + control.width;
    return columns;
  }

  render() {
    this.lastColumn = 0;
    this.lastRow = 0;

    let row: { key: number; values: Corpix.Entities.FormElement[] };
    const rows = groupByArray(this.props.form.elements, 'row');

    return (
      <Grid>
        <For each="row" of={rows}>
          <Grid.Row key={row.key}>{row.values.map(element => this.renderColumn(element))}</Grid.Row>
        </For>
      </Grid>
    );
  }
}
