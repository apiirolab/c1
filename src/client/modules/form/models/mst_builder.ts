import { Instance, types } from 'mobx-state-tree';
import { UndoManager } from 'mst-middlewares';

import { DataSetNode } from './dataset_model';
import { FormStore } from './form_store';
import { safeEval } from './form_utils';

let time = Date.now();
let i = 0;
function shortId() {
  return (time + i++).toString();
}

function mstTypeFactory(desc: DataSetNode): any {
  switch (desc.type) {
    case 'id':
      return types.optional(types.identifier, shortId);
    case 'expression':
      return undefined;
    case 'array':
      return types.optional(
        types.array(types.optional(mstTypeFactory(desc.items), desc.items.defaultValue)),
        desc.default || []
      );
    case 'string':
      return types.optional(types.union(types.string, types.undefined), desc.default);
    case 'integer':
      return types.optional(types.union(types.number, types.string, types.undefined), desc.default);
    case 'number':
      return types.optional(types.union(types.number, types.string, types.undefined), desc.default);
    case 'date':
      return types.optional(types.union(types.Date, types.string, types.undefined), desc.default);
    case 'boolean':
      return types.optional(
        types.union(types.boolean, types.string, types.undefined),
        desc.default
      );
    case 'object':
      return types.optional(buildStore(desc), {});
    case undefined:
      return types.string;
  }
  throw new Error('MST Type not supported: ' + desc.type);
}

export function buildStore(store: DataSetNode) {
  // prepare model and views

  const mstDefinition: { [index: string]: any } = {};

  /* =========================================================
    EXPRESSIONS
    ======================================================== */
  const properties = Object.getOwnPropertyNames(store.properties);

  const viewDefinition = () => {
    const view = {};

    for (let key of properties) {
      let node = store.properties[key];
      // expressions do not need state tree entry they are evaluated automatically
      if (node.type === 'expression') {
        (view as any).__defineGetter__(key, function() {
          // @ts-ignore
          return safeEval(this, node.default);
        });
      }
    }

    return view;
  };

  /* =========================================================
      MST Nodes
     ======================================================== */

  for (let key of properties) {
    let node = store.properties[key];
    let definition = mstTypeFactory(node);
    if (definition) {
      mstDefinition[key] = types.maybe(definition);
    }
  }

  if (store.parent == null) {
    mstDefinition.history = types.optional(UndoManager, {});
  }

  // build tree
  const mst = FormStore.named('FormStore')
    .props(mstDefinition)
    .views(viewDefinition)
    .actions(() => ({
      getSchema(key: string) {
        return key ? store.properties[key] : store;
      }
    }));

  return mst;
}
