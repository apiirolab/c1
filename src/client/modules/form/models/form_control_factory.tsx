import * as React from 'react';

import { Button } from 'semantic-ui-react';
import { DataSet } from './form_model';
import { CheckboxView } from '../views/checkbox_view';
import { FormulaView } from '../views/formula_view';
import { InputView } from '../views/input_view';
import { RadioView } from '../views/radio_view';
import { RepeaterView } from '../views/repeater_view';
import { SelectView } from '../views/select_view';

export function renderControl(
  control: Corpix.Entities.FormElement,
  dataSet: DataSet,
  handlers: { [index: string]: () => void }
) {
  const formElement = control as Corpix.Entities.FormControl;

  if (dataSet.isExpression(formElement.source)) {
    return <FormulaView owner={dataSet} formControl={formElement} />;
  }

  switch (formElement.control) {
    case 'input':
      return <InputView owner={dataSet} formControl={formElement} />;
    case 'select':
      return <SelectView owner={dataSet} formControl={formElement} />;
    case 'checkbox':
      return <CheckboxView owner={dataSet} formControl={formElement} />;
    case 'radio':
      return <RadioView owner={dataSet} formControl={formElement} />;
    case 'repeater':
      return <RepeaterView owner={dataSet} formControl={formElement} />;
    case 'deleteButton':
      return <Button icon="trash" color="red" onClick={handlers.delete} />;
  }

  throw new Error('Not implemented: ' + formElement.control);
}
