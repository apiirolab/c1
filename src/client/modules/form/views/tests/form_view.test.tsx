import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Segment } from 'semantic-ui-react';

import { create } from 'shared/test_data';
import { FormView } from '../form_view';
import { FormModel } from '../../models/form_model';

import DevTools from 'mobx-react-devtools';

describe('Form', () => {
  const descriptors = [
    create.descriptorDao({ name: 'owner.personal.name' }),
    create.descriptorDao({ name: 'owner.personal.age', type: 'int' }),
    create.descriptorDao({
      name: 'younger',
      type: 'int',
      expression: `self['owner.personal.age'] - 10`
    }),
    create.descriptorDao({
      name: 'older',
      type: 'int',
      expression: `self['owner.personal.age'] + 10`
    })
  ];

  const data = [
    { name: 'owner.personal.name', value: 'Tomas' },
    {
      name: 'owner.personal.age',
      value: 33
    }
  ];

  const dataSet = FormModel.buildMST(descriptors, data);

  storyOf(
    'Viewer',
    {
      get component() {
        const form = new FormModel(
          create.formDao({
            elements: [
              {
                id: '1',
                row: 0,
                column: 0,
                width: 16,
                control: 'input',
                label: 'Mimo',
                inputLabel: 'http://',
                source: 'owner.personal.name'
              },
              {
                id: '2',
                row: 1,
                column: 0,
                width: 8,
                control: 'input',
                source: 'owner.personal.age',
                label: 'Age'
              },
              { id: '3', row: 2, column: 2, width: 6, control: 'input' },
              { id: '4', row: 2, column: 8, width: 6, control: 'input' },
              { id: '5', row: 3, column: 2, width: 6, control: 'input' }
            ]
          })
        );

        // just another notation
        return (
          <Segment className="ui form">
            <DevTools />
            <FormView form={form} data={dataSet} />
          </Segment>
        );
      }
    },
    data => {
      it('renders correctly', () => {
        const component = renderer.create(data.component);
        expect(component).toMatchSnapshot();
      });
    }
  );
});
