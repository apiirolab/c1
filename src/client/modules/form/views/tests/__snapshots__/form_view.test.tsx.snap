// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Viewer renders correctly 1`] = `
<div
  className="ui segment ui form"
>
  <div>
    <div>
      <div
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "rgba(0, 0, 0, 0.1)",
            "borderRadius": "0 0 2px 2px",
            "borderStyle": "solid",
            "borderWidth": "0 1px 1px",
            "color": "rgba(0, 0, 0, 0.8)",
            "display": "flex",
            "fontFamily": "Helvetica, sans-serif",
            "height": "26px",
            "padding": "0 5px",
            "position": "fixed",
            "right": "20px",
            "top": "0px",
            "zIndex": "65000",
          }
        }
      >
        <button
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "active": Object {
                ":hover": Object {
                  "opacity": 1,
                },
                "opacity": 1,
              },
              "background": "transparent none center / 16px 16px no-repeat",
              "backgroundImage": "url(data:image/svg+xml;base64,PHN2ZyBiYXNlUHJvZmlsZT0iYmFzaWMiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmlld0JveD0iMCAwIDE2IDE2IiBoZWlnaHQ9IjE2IiB3aWR0aD0iMTYiPgogICAgPGcgc3Ryb2tlPSIjMDAwIiBmaWxsPSJub25lIiBzdHJva2Utd2lkdGg9IjEiPgogICAgICAgIDxjaXJjbGUgY3g9IjguNSIgY3k9IjguNSIgcj0iNiIvPgogICAgICAgIDxwYXRoIGQ9Ik04LjUgMTBWNCIvPgogICAgPC9nPgogICAgPGcgc3Ryb2tlPSJub25lIiBmaWxsPSIjMDAwIj4KICAgICAgICA8Y2lyY2xlIGN4PSI4LjUiIGN5PSI4LjUiIHI9IjEiLz4KICAgICAgICA8cGF0aCBkPSJNNy41IDFoMnYxLjVoLTJ6Ii8+CiAgICAgICAgPHBhdGggZD0iTTE0IDEuNmwtLjcuNy4zLjMtLjcuOC43LjcuOC0uNy4zLjMuNy0uN3oiLz4KICAgIDwvZz4KPC9zdmc+Cg==)",
              "border": "none",
              "cursor": "pointer",
              "hover": Object {
                "opacity": 0.7,
              },
              "margin": "0 10px",
              "opacity": 0.45,
              "width": "26px",
            }
          }
          title="Visualize component re-renders"
          type="button"
        />
        <button
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "active": Object {
                ":hover": Object {
                  "opacity": 1,
                },
                "opacity": 1,
              },
              "background": "transparent none center / 16px 16px no-repeat",
              "backgroundImage": "url(data:image/svg+xml;base64,PHN2ZyBiYXNlUHJvZmlsZT0iYmFzaWMiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmlld0JveD0iMCAwIDE2IDE2IiBoZWlnaHQ9IjE2IiB3aWR0aD0iMTYiPgogICAgPGcgc3Ryb2tlPSIjMDAwIiBmaWxsPSJub25lIiBzdHJva2Utd2lkdGg9IjEiPgogICAgICAgIDxwYXRoIGQ9Ik0yLjUgMi41aDl2MmgtOXoiLz4KICAgICAgICA8cGF0aCBkPSJNNy41IDcuNWg3djJoLTd6Ii8+CiAgICAgICAgPHBhdGggZD0iTTcuNSAxMi41aDd2MmgtN3oiLz4KICAgICAgICA8cGF0aCBkPSJNNC41IDQuNXY5aDMiLz4KICAgICAgICA8cGF0aCBkPSJNNy41IDguNWgtMyIvPgogICAgPC9nPgo8L3N2Zz4K)",
              "border": "none",
              "cursor": "pointer",
              "hover": Object {
                "opacity": 0.7,
              },
              "margin": "0 10px",
              "opacity": 0.45,
              "width": "26px",
            }
          }
          title="Select a component and show its dependency tree"
          type="button"
        />
        <button
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "active": Object {
                ":hover": Object {
                  "opacity": 1,
                },
                "opacity": 1,
              },
              "background": "transparent none center / 16px 16px no-repeat",
              "backgroundImage": "url(data:image/svg+xml;base64,PHN2ZyBiYXNlUHJvZmlsZT0iYmFzaWMiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmlld0JveD0iMCAwIDE2IDE2IiBoZWlnaHQ9IjE2IiB3aWR0aD0iMTYiPgogICAgPGcgc3Ryb2tlPSIjMDAwIiBmaWxsPSJub25lIiBzdHJva2Utd2lkdGg9IjEiPgogICAgICAgIDxwYXRoIGQ9Ik0xMi41IDMuNWgtOGMtMS4xIDAtMiAuOS0yIDJ2NWMwIDEuMS45IDIgMiAyaDF2Mmw1LTJoMmMxLjEgMCAyLS45IDItMnYtNWMwLTEuMS0uOS0yLTItMnoiLz4KICAgICAgICA8cGF0aCBkPSJNNSA2LjVoNyIvPgogICAgICAgIDxwYXRoIGQ9Ik01IDkuNWg3Ii8+CiAgICA8L2c+Cjwvc3ZnPgo=)",
              "border": "none",
              "cursor": "pointer",
              "hover": Object {
                "opacity": 0.7,
              },
              "margin": "0 10px",
              "opacity": 0.45,
              "width": "26px",
            }
          }
          title="Log all MobX state changes and reactions to the browser console (use F12 to show / hide the console). Use Chrome / Chromium for an optimal experience"
          type="button"
        />
      </div>
    </div>
    <div />
  </div>
  <div
    className="ui grid"
  >
    <div
      className="row"
    >
      <div
        className="six wide column"
      >
        <div
          className="field"
        >
          <label>
            Mimo
          </label>
          <div
            className="ui labeled input"
          >
            <div
              className="ui label label"
              onClick={[Function]}
            >
              http://
            </div>
            <input
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="two wide column"
      />
      <div
        className="six wide column"
      >
        <div
          className="field"
        >
          <div
            className="ui input"
          >
            <input
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="two wide column"
      />
      <div
        className="six wide column"
      >
        <div
          className="field"
        >
          <div
            className="ui input"
          >
            <input
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="six wide column"
      >
        <div
          className="field"
        >
          <div
            className="ui input"
          >
            <input
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="two wide column"
      />
      <div
        className="six wide column"
      >
        <div
          className="field"
        >
          <div
            className="ui input"
          >
            <input
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
